webui/react/src/e2e/models/BaseComponent.ts

```diff
@@ -0,0 +1,83 @@
+import { type Locator } from '@playwright/test';
+import { BasePage } from './BasePage';
+
+/**
+ * Used in the constructor for BaseComponent
+ */
+export interface BaseComponentProps {
+  parent: BasePage | BaseComponent
+  selector?: string
+}
+
+/**
+ * Returns the representation of a Component.
+ * 
+ * @remarks
+ * This constructor is a base class for any component in src/components/.
+ * 
+ * @param {Object} obj
+ * @param {BasePage | BaseComponent} obj.parent - The parent used to locate this BaseComponent
+ * @param {string} [obj.selector] - Used instead of `defaultSelector`
+ */
+export class BaseComponent {
+  readonly defaultSelector: undefined | string;
+
+  readonly #selector: string;
+  protected readonly _parent: BasePage | BaseComponent;
+  protected _locator: Locator | undefined;
+
+  constructor({ parent, selector }: BaseComponentProps) {
+    if (typeof this !== typeof BaseComponent && typeof this.defaultSelector === 'undefined') {
+      throw new Error(`defaultSelector is undefined in class ${typeof this}`);
+    }
+    if (typeof this === typeof BaseComponent && typeof selector === 'undefined') {
+      throw new Error(`BaseComponent needs a selector`);
+    }
+
+    // guardrails above ensure that either selector or defaultSelector are defined
+    this.#selector = selector || this.defaultSelector!;
+    this._parent = parent;
+  }
+
+  /**
+   * The playwright locator method from this model's locator
+   */
+  get pwLocatorFunction() { return this.locateSelf.locator }
+
+  /**
+   * The playwright Locator that represents this model
+   */
+  get locateSelf(): Locator {
+    if (typeof this._locator === 'undefined') {
+      this._locator = this._parent.pwLocatorFunction(this.#selector);
+      Object.freeze(this._locator)
```

Comment:
Why are we freezing the object here? It makes me a little nervous. Hopefully playwright doesn't keep any mutable data on Locator classes under the hood!

Reply:
I wanted the attribute to only be instantiated once. it's kind of like saying
```
const myLocator = page.locator('my_selector')
```
I didn't want to create the "const" until we started using it. it behaves like a `readonly` that we're allowed to write to once.